### PR TITLE
[wip] Update to pyodide 0.22.0, use pyodide and serviceworker types

### LINKED ISSUE #792: notebook `attachment:` URLs incorrectly routed to ServiceWorker
## Description

When viewing a notebook with attachments (such as images), the attachments aren't shown, and the console  shows the following errors:

```
GET attachment:16109f49-1053-42ab-b02d-9685e0743e6e.png net::ERR_UNKNOWN_URL_SCHEME
attachment:16109f49-1053-42ab-b02d-9685e0743e6e.png?1662552778797:1          
GET attachment:16109f49-1053-42ab-b02d-9685e0743e6e.png?1662552778797 net::ERR_UNKNOWN_URL_SCHEME
```

## Reproduce

- go to https://jupyterlite.rtfd.io/en/latest/try/lab?path=pyolite%2Fipycytoscape.ipynb
- open the browser console
- see errors
- don't see images

## Expected behavior

- see images

## Context

Probably after #686, so [0.1.0b10](https://github.com/jupyterlite/jupyterlite/blob/main/CHANGELOG.md#010b10)

### PR DESCRIPTION
**Pull request by bollwyvl**
## References

- could be moved to new repo if #854 lands first
- fixes #792 (it wasn't really broken)
- investigates #751

## Code changes

- [x] add exact-pinned `pyodide` to `pyodide-kernel/package.json#/devDependencies`
  - only ever `import type`d in `.ts`
  - is the source of truth for all other versions
- [x] bump to [`pyodide 0.22.0a3`](https://github.com/pyodide/pyodide/blob/0.22.0a3/docs/project/changelog.md#unreleased)
- [x] remove duplicated `ERRNO_CODES`
- [x] remove use of private `pyodide._module`
- [x] update key dev deps
  - [x] `lerna 6`
  - [x] `typedoc`
  - [x] `typescript`
  - [x] `prettier`    
- [x] refactor `BroadcastChannel` out of server, put in contents
- [x] make pyolite more resilient to missing `ServiceWorker` and `BroadcastChannel` 
- [x] moves `services.js` into `server/sw.ts`
  - [x] add a new webpack plugin to hoist it up to the root of the app (or the scope is wrong) and add the content hash to the file name
- [x] add `SKIP_ESLINT` env var
  - > the pre-commit hooks still fail sometimes, so `git commit -n` it shall be!

## User-facing changes

- potentially better error messages on startup if Things Go Wrong

## Backwards-incompatible changes

- n/a
  - the `sqlite3` change might have been surprising...
    - but we'll always be loading it so `IPython` can load

### CODE REVIEW

File: packages/pyolite-kernel/src/worker.ts [hunk 1]
```diff
@@ -582,10 +462,10 @@ export class PyoliteRemoteKernel {
     reject: () => void;
     resolve: () => void;
   } | null = null;
+  protected _pyodide: Pyodide.PyodideInterface = null as any;
```
Comment (bollwyvl) [inline packages/pyolite-kernel/src/worker.ts:465]: not proud of this...

File: packages/pyolite-kernel/src/worker.ts [hunk 2]
```diff
@@ -582,10 +462,10 @@ export class PyoliteRemoteKernel {
...
   /** TODO: real typing */
   protected _localPath = '';
   protected _driveName = '';
-  protected _pyodide: any;
   protected _kernel: any;
```
Comment (bollwyvl) [inline packages/pyolite-kernel/src/worker.ts:469]: These `any`s still kill me. Among the tasks we should explore in #386 are:
- doing a thorough typing of all our custom classes
- generating typescript types from the python types https://github.com/pyodide/pyodide/issues/2162

File: packages/server/src/serviceworker.ts [hunk 3]
```diff
@@ -28,35 +29,44 @@ export class ServiceWorkerRegistrationWrapper
...
+    const workerUrl = URLExt.join(PageConfig.getBaseUrl(), 'services.js');
+    let registration: ServiceWorkerRegistration | null = null;
+
+    if (!serviceWorker) {
+      console.warn('ServiceWorkers not supported in this browser');
+    } else if (serviceWorker.controller) {
+      registration =
+        (await serviceWorker.getRegistration(serviceWorker.controller.scriptURL)) ||
+        null;
+      console.info('JupyterLite ServiceWorker was already registered');
```
Comment (bollwyvl) [inline packages/server/src/serviceworker.ts:47]: ref #751

**Comment by bollwyvl** [comment]
Seems to be working on CI. Still no progress on #840, even when trying a more complex setup for handling potential race conditions on the `BroadcastChannel`.

Had a look at a few other things, rolled them back:
- [workbox](https://github.com/GoogleChrome/workbox) for serviceworker
  - there's a lot here, like plugins and strategies, but I don't know how we'd get them to work inside of a simple script
  - we _can_ however pass some config in via `sw-1234.js?param=value`, so there might be some things that would work well there in a more configurable way
- bringing back `importhook` #254
  - causes IPython to fail loading due to `pygments` tricksiness, so probably would break other things

File: packages/contents/src/drivefs.ts [hunk 4]
```diff
@@ -177,16 +193,16 @@ export class DriveFSEmscriptenStreamOps implements IEmscriptenStreamOps {
     length: number,
     position: number
   ): number {
-    if (length <= 0 || stream.file === undefined) {
+    if (
+      length <= 0 ||
+      stream.file === undefined ||
+      position >= (stream.file.data.length || 0)
```
Comment (bollwyvl) [inline packages/contents/src/drivefs.ts:199]: ref
- https://github.com/jupyterlite/jupyterlite/issues/840#issuecomment-1313061817 
- https://github.com/jupyterlite/jupyterlite/commit/c7c29078edf28f1828023385fde8e4d130a4630a#r89808665

There's still _something_ happening in the pyolite case, where doing successive `writes` without closing between them yields a `database disk image is malformed`

File: packages/pyolite-kernel/src/worker.ts [hunk 5]
```diff
@@ -197,6 +78,7 @@ export class PyoliteRemoteKernel {
   protected async initKernel(options: IPyoliteWorkerKernel.IOptions): Promise<void> {
     // from this point forward, only use piplite (but not %pip)
     await this._pyodide.runPythonAsync(`
+      await piplite.install(['sqlite3'], keep_going=True)
```
Comment (bollwyvl) [inline packages/pyolite-kernel/src/worker.ts:81]: This will still be needed in #909

File: dodo.py [hunk 6]
```diff
@@ -1073,23 +1063,38 @@ class P:
...
     PACKAGE_JSONS = {
-        p.parent.name: json.loads(p.read_text(**C.ENC)) for p in P.PACKAGE_JSONS
+        parent: json.loads(p.read_text(**C.ENC))
+        for parent, p in P.PACKAGE_JSONS.items()
     }
+    PYODIDE_JS_VERSION = PACKAGE_JSONS["pyolite-kernel"]["devDependencies"]["pyodide"]
+    PYODIDE_CDN_URL = APP_SCHEMA_PYOLITE["properties"]["pyodideUrl"]["default"]
+    PYODIDE_VERSION = _js_version_to_py_version(PYODIDE_JS_VERSION)
+    PYODIDE_JS = PYODIDE_CDN_URL.split("/")[-1]
+    PYODIDE_ARCHIVE = f"pyodide-{PYODIDE_VERSION}.tar.bz2"
```
Comment (bollwyvl) [inline dodo.py:1092]: this name will need to change, removing `-build` in #909

Comment (jtpio) [reply dodo.py:1092]: Sounds good let's keep track of that in #909.

**Pull request merged**